lib/systems/platforms.nix

```diff
@@ -482,6 +482,55 @@ rec {
     };
   };
 
+  # MIPS ABI table transcribed from here: https://wiki.debian.org/Multiarch/Tuples
+
+  # can execute on 32bit chip
+  mips-linux-gnu                = { gcc = { arch = "mips32r2"; abi = "o32"; }; };
+  mipsel-linux-gnu              = { gcc = { arch = "mips32r2"; abi = "o32"; }; };
+  mipsisa32r6-linux-gnu         = { gcc = { arch = "mips32r6"; abi = "o32"; }; };
+  mipsisa32r6el-linux-gnu       = { gcc = { arch = "mips32r6"; abi = "o32"; }; };
+
+  # require 64bit chip (for more registers, 64-bit floating point, 64-bit "long long") but use 32bit pointers
+  mips64-linux-gnuabin32        = { gcc = { arch = "mips64r2"; abi = "n32"; }; };
+  mips64el-linux-gnuabin32      = { gcc = { arch = "mips64r2"; abi = "n32"; }; };
+  mipsisa64r6-linux-gnuabin32   = { gcc = { arch = "mips64r6"; abi = "n32"; }; };
+  mipsisa64r6el-linux-gnuabin32 = { gcc = { arch = "mips64r6"; abi = "n32"; }; };
+
+  # 64bit pointers
+  mips64-linux-gnuabi64         = { gcc = { arch = "mips64r2"; abi =  "64"; }; };
+  mips64el-linux-gnuabi64       = { gcc = { arch = "mips64r2"; abi =  "64"; }; };
+  mipsisa64r6-linux-gnuabi64    = { gcc = { arch = "mips64r6"; abi =  "64"; }; };
+  mipsisa64r6el-linux-gnuabi64  = { gcc = { arch = "mips64r6"; abi =  "64"; }; };
```

Comment:
Yes so these *names* are better for `exmaples.nix`. These should be deduped, and given names without "linux" or "gnu vs musl", since that is no longer what they are about.

BTW see also https://github.com/NixOS/nixpkgs/blob/master/lib/systems/architectures.nix. That file is sadly not really engaged with everything else very much --- it should be incorperated better, someday. But I thought I would bring it up since this fine-grained, predictable, control over such specifics is somewhat reminiscent of it.

Reply:
Okay.  I am still really confused about `platforms.nix` vs `examples.nix`.  The values in examples.nix aren't just examples -- in fact it looks like examples.nix is really the official list of platforms.  When you `nix-env -iA pkgsCross.FOO` it will only work if `FOO` is in `examples.nix`.  It doesn't look in `platforms.nix`!

So this leaves me very confused about what is the point of `platforms.nix`.  Why aren't `platforms.nix` and `examples.nix` merged into a single file?

Reply:
For bi-endian processors these `gcc = { ... }` specifications are ambiguous.  GCC needs to be told which endian (big or little) it is generating code for.  Currently it is getting that information from the `config`, so it's not really clear to me what the point is of having these ambiguous partially-specified-gcc-code-generation modes floating around.

A lot of this stuff really needs to be cleaned up and organized better, but I don't want to delay this getting merged because of that.

> BTW see also https://github.com/NixOS/nixpkgs/blob/master/lib/systems/architectures.nix.  That file is sadly not really engaged with everything else very much

That's probably okay... this table of "CPU feature bits" is just a compiler customization.  Changing these flags won't break your ability to link dynamically (changing the ABI does this) nor will it break your ability to use a kernel that was compiled differently (changing endianness does this).  You can compile your program, its libc, and the kernel all with completely different `march=` settings and they will work together perfectly.

Also, the names in that table are kind of arbitrarily chosen by GCC.  LLVM tries to imitate them where it can (for compatibility) but there are lots of cases missing (for example LLVM chokes on `-mcpu=cortex-a72.cortex-a53` for ARM "big.LITTLE" devices like RK3399).

**GNU tuples really do capture exactly what is needed in order to make sure a program, libc, and kernel can work together -- that, and nothing more.  I really think that nixpkgs would be simplified and run into fewer problems if it treated "platform" as meaning exactly the same thing as "fully-specified GNU 5-tuple".  Remember, GNU tuples have *five* fields, not four.  The characters "abi" are a field-separator (yeah that was a dumb choice).**

Reply:
Thanks for the info @a-m-joseph, and interesting historical note there from Debian! I myself rewrote much of GNU Config, so yes I think there is good hope for being able reconcile such things in the future.